From 086fff4e25056e05350763d09733c39ace114e65 Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Wed, 20 Aug 2014 18:58:46 -0400
Subject: [scsi] sd: Do not call do_div() with a 64-bit divisor

Message-id: <1408561176-18133-8-git-send-email-emilne@redhat.com>
Patchwork-id: 88428
O-Subject: [RHEL7.1 PATCH e-stor 07/57] [SCSI] sd: Do not call do_div() with a 64-bit divisor
Bugzilla: 1132092
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Ewan D. Milne" <emilne@redhat.com>

upstream commit ef80d1e18b014af08741cf688e3fdda1fb71363f

Tested by Ewan D. Milne and Barry Donahue with a combination
of manual and automated testing.  No issues seen.

upstream commit ef80d1e18b014af08741cf688e3fdda1fb71363f
Author: Geert Uytterhoeven <geert@linux-m68k.org>
Date:   Mon Nov 4 10:21:05 2013 +0100

    [SCSI] sd: Do not call do_div() with a 64-bit divisor

    do_div() is meant for divisions of 64-bit number by 32-bit numbers.
    Passing 64-bit divisor types caused issues in the past on 32-bit platforms,
    cfr. commit ea077b1b96e073eac5c3c5590529e964767fc5f7 ("m68k: Truncate base
    in do_div()").

    As scsi_device.sector_size is unsigned (int), factor should be unsigned
    int, too.

    Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>
    Signed-off-by: James Bottomley <JBottomley@Parallels.com>

Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/sd.c b/drivers/scsi/sd.c
index 35bc29e..869cb2a 100644
--- a/drivers/scsi/sd.c
+++ b/drivers/scsi/sd.c
@@ -1595,7 +1595,7 @@ static unsigned int sd_completed_bytes(struct scsi_cmnd *scmd)
   end_lba <<= 1;
  } else {
   /* be careful ... don't want any overflows */
-  u64 factor = scmd->device->sector_size / 512;
+  unsigned int factor = scmd->device->sector_size / 512;
   do_div(start_lba, factor);
   do_div(end_lba, factor);
  }
-- 
1.7.1